### PR TITLE
API for users, registration, reset password and facebook

### PR DESCRIPTION
**Pull request by IraVolyk**
Test **AuthenticationController::login**
image

Test **AuthenticationController::logout**
image

Test - **password::new** for send email with token to reset password
image

Test **password::edit** (reset password)
image

Generate token for Facebook
image

Test **FacebookAuthenticationController::create**
image

### CODE REVIEW

File: app/views/api/v1/users/index.json.jbuilder [hunk 1]
```diff
@@ -0,0 +1,5 @@
+json.array! users do |user|
+  json.id user.id
```
Comment (yuravenzhik) [inline app/views/api/v1/users/index.json.jbuilder:2]: partial

File: app/views/api/v1/authentication/login.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,6 @@
+json.user do |json|
+  json.id user.id
+  json.email user.email
+  json.full_name user.full_name
+end
+json.token jwt
```
Comment (yuravenzhik) [inline app/views/api/v1/authentication/login.json.jbuilder:6]: place token to header

File: app/models/user.rb [hunk 3]
```diff
@@ -57,11 +57,31 @@ def full_name
...
+    data = {
+      email: profile['email'],
+      first_name: profile['first_name'],
+      last_name: profile['last_name'],
+      uid: profile['id'],
+      provider: 'facebook',
+      password: SecureRandom.urlsafe_base64
+    }
+ 
+    if user = User.find_by(uid: data['uid'], provider: 'facebook')
```
Comment (yuravenzhik) [inline app/models/user.rb:73]: if (user = User.find_by(uid: data['uid'], provider: 'facebook'))

File: app/models/user.rb [hunk 4]
```diff
@@ -57,11 +57,31 @@ def full_name
     where("concat_ws('OR', LOWER(email), LOWER(first_name), LOWER(last_name)) LIKE LOWER(?)", "%#{user}%")
   }
 
+  def self.find_or_create_with_facebook_access_token(oauth_access_token)
+    graph = Koala::Facebook::API.new(oauth_access_token)
```
Comment (yuravenzhik) [inline app/models/user.rb:61]: put this method to service object

File: app/mailers/password_mailer.rb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+class PasswordMailer < ApplicationMailer
+  before_action { @user = params[:user] }
+  default from: "not-replay@projector.com"
```
Comment (yuravenzhik) [inline app/mailers/password_mailer.rb:3]: put to action mailer config

File: app/controllers/api/v1/base_controller.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+          begin
+            JWT.decode(token, Rails.application.secrets.secret_key_base, true, algorithm: 'HS256')
+          rescue JWT::DecodeError
+            []
+          end
+        end
+      end
+
+      def session_user
+        decoded_hash = decoded_token
```
Comment (yuravenzhik) [inline app/controllers/api/v1/base_controller.rb:28]: service

File: app/controllers/api/v1/authentication_controller.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+module Api
+  module V1
+    class AuthenticationController < BaseController
+      skip_before_action :require_login, only: %i[login auto_login]
+
+      def login
+        user = User.find_by(email: params[:email])
+
+        if user && user.valid_password?(params[:password])
+          token =  Users::CreateTokenService.call(user)
```
Comment (yuravenzhik) [inline app/controllers/api/v1/authentication_controller.rb:10]: duplicated code

File: app/controllers/api/v1/authentication_controller.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+          render :login, locals: { user: user, jwt: token}, status: 201
+        else
+          render json: { failure: 'Log in failed! Username or password invalid!' }, status: 401
+        end
+      end
+
+      def auto_login
+        if session_user
+          render json: session_user
```
Comment (yuravenzhik) [inline app/controllers/api/v1/authentication_controller.rb:19]: jbuilder

File: app/controllers/api/v1/password_controller.rb [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+        PasswordMailer.with(user: @user).reset_password.deliver_later
+        render json: { success: 'Successfully sended email.' }, status: :not_acceptable
+      end
+
+      def edit
+        user_token = params["reset_password_token"]
+        @user = User.find_by(reset_password_token: params[:reset_password_token])
+
+        if @user.nil?
+          render json: { errors: "User not found!" }, status: :not_acceptable
```
Comment (yuravenzhik) [inline app/controllers/api/v1/password_controller.rb:20]: 1. 404
2. find_by! + handle ActiveRecord::RecordNotFound

File: app/controllers/api/v1/password_controller.rb [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+          return
+        end
+        
+        if @user.reset_password_token != user_token
```
Comment (yuravenzhik) [inline app/controllers/api/v1/password_controller.rb:24]: лишняя логика

File: app/controllers/api/v1/password_controller.rb [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+          render json: { errors: "Token to reset passwor in not valid." }, status: :not_acceptable
+          return
+        end
+
+        if Time.now.to_i > @user.reset_password_sent_at.to_i
```
Comment (yuravenzhik) [inline app/controllers/api/v1/password_controller.rb:29]: to scope `:with_active_reset_password`

File: app/controllers/api/v1/base_controller.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+
+      def session_user
+        user = Users::DecryptTokenService.call(auth_header)
+      end
+
+      def logged_in?
+        session_user.present?
+      end
+
+      def require_login
```
Comment (yuravenzhik) [inline app/controllers/api/v1/base_controller.rb:22]: authenticate!

File: app/controllers/api/v1/facebook_authentications_controller.rb [hunk 13]
```diff
@@ -0,0 +1,25 @@
+require 'koala'
+module Api
+  module V1
+    class FacebookAuthenticationsController < BaseController
+
+      skip_before_action :require_login
+
+      respond_to :json
```
Comment (yuravenzhik) [inline app/controllers/api/v1/facebook_authentications_controller.rb:8]: !

File: app/controllers/api/v1/authentication_controller.rb [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+      def auto_login
+        if session_user
+          render partial: 'api/v1/users/user', locals: { user: user }, status: 200
+        else
+          render json: { error: 'User is not Logged In.' }, status: 401
+        end
+      end
+
+      def user_is_authed
```
Comment (yuravenzhik) [inline app/controllers/api/v1/authentication_controller.rb:26]: -

File: app/services/users/find_or_create_with_facebook_access_token.rb [hunk 15]
```diff
@@ -0,0 +1,30 @@
+module Users
+  class FindOrCreateWithFacebookAccessToken < ApplicationService
+    attr_accessor :oauth_access_token
+
+    def initialize(oauth_access_token)
+      @oauth_access_token = oauth_access_token
+    end
+  
+    def call
+      graph = Koala::Facebook::API.new(oauth_access_token)
```
Comment (yuravenzhik) [inline app/services/users/find_or_create_with_facebook_access_token.rb:10]: 1

File: app/services/users/find_or_create_with_facebook_access_token.rb [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+      profile = graph.get_object('me', fields: ['first_name', 'last_name', 'email'])
+  
+      data = {
+        email: profile['email'],
+        first_name: profile['first_name'],
+        last_name: profile['last_name'],
+        uid: profile['id'],
+        provider: 'facebook',
+        password: SecureRandom.urlsafe_base64
+      }
```
Comment (yuravenzhik) [inline app/services/users/find_or_create_with_facebook_access_token.rb:20]: 2

File: app/services/users/find_or_create_with_facebook_access_token.rb [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+  
+      if (user = User.find_by(uid: data['uid'], provider: 'facebook'))
+        user.update_attributes(data)
+      else
+        User.create(data)
+      end
```
Comment (yuravenzhik) [inline app/services/users/find_or_create_with_facebook_access_token.rb:26]: 3

File: spec/requests/api/v1/users_request_spec.rb [hunk 18]
```diff
@@ -0,0 +1,163 @@
...
+    end
+  end
+
+  describe "GET api/v1/users#index" do
+    context "correct params are passed" do
+      let!(:user2) { create(:user) }
+      subject { get api_v1_users_path( format: :json), headers: token  }
+      it "return json wirth 2 users" do
+        subject
+        expect(JSON.parse(response.body)).to include_json([
```
Comment (yuravenzhik) [inline spec/requests/api/v1/users_request_spec.rb:128]: to_eq

File: app/controllers/api/v1/users_controller.rb [hunk 19]
```diff
@@ -0,0 +1,33 @@
...
+      def index
+        users = User.search(params[:search]).limit(10)
+        render :index, locals: { users: users }
+      end
+
+      def create
+        user = User.new(user_params)
+        if user.save
+          generate_token_and_set_to_header(user)
+          render partial: 'api/v1/users/user', locals: { user: user }, status: 201
```
Comment (etherell) [inline app/controllers/api/v1/users_controller.rb:15]: Why not render show?

File: app/controllers/api/v1/users_controller.rb [hunk 20]
```diff
@@ -0,0 +1,33 @@
+module Api
+  module V1
+    class UsersController < BaseController
+      skip_before_action :authenticate!, only: %i[create]
+
+      def index
+        users = User.search(params[:search]).limit(10)
```
Comment (etherell) [inline app/controllers/api/v1/users_controller.rb:7]: I think users should be with "@"

File: app/controllers/api/v1/users_controller.rb [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+        if user.save
+          generate_token_and_set_to_header(user)
+          render partial: 'api/v1/users/user', locals: { user: user }, status: 201
+        else
+          render json: { errors: user.errors.full_messages }, status: 422
+        end
+      end
+
+      def show
+        render :show, locals: { user: session_user }, status: 200
```
Comment (etherell) [inline app/controllers/api/v1/users_controller.rb:22]: @user = session_user

**Pull request merged**